graph-commons/src/test/scala/io/renku/http/client/GitLabClientSpec.scala

```diff
@@ -0,0 +1,191 @@
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "fetch commits using the given oauth token" in new TestCase {
+      val authAccessToken = oauthAccessTokens.generateOne
+
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .withHeader("PRIVATE-TOKEN", equalTo(authAccessToken.value))
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+    }
+
+    "Handle empty JSON" in new TestCase {
+      override val result: String = "{}"
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .willReturn(
+            okJson(result)
+              .withHeader("X-Next-Page", maybeNextPage.map(_.show).getOrElse(""))
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(result)).unsafeRunSync() shouldBe result
+
+    }
+
+    "Handle 404 NOT FOUND without throwing" in new TestCase {
+      stubFor {
+        callMethod(method, s"api/v4/$path")
+          .willReturn(
+            notFound()
+          )
+      }
+
+      client.send(method, path, endpointName)(_ => IO(())).unsafeRunSync() shouldBe ()
+
+    }
+
+    "return an UnauthorizedException if remote client responds with UNAUTHORIZED" in new TestCase {
+      stubFor {
+        callMethod(method, s"/api/v4/$path")
+          .willReturn(
+            unauthorized()
+          )
+      }
+
+      private implicit lazy val mapResponse: PartialFunction[(Status, Request[IO], Response[IO]), IO[Option[Int]]] = {
+        case (Ok, _, response)    => response.as[List[Int]].map(_.headOption)
+        case (NotFound, _, _)     => Option.empty[Int].pure[IO]
+        case (Unauthorized, _, _) => UnauthorizedException.raiseError[IO, Option[Int]]
+      }
+
+      intercept[Exception] {
+        client.send(method, path, endpointName)(mapResponse).unsafeRunSync()
+      } shouldBe UnauthorizedException
+    }
+
+    "return an Exception if remote client responds with status neither OK nor UNAUTHORIZED" in new TestCase {
+      stubFor {
+        callMethod(method, s"/api/v4/$path")
+          .willReturn(
+            badRequest()
+          )
+      }
+
+      private implicit lazy val mapResponse: PartialFunction[(Status, Request[IO], Response[IO]), IO[Option[Int]]] = {
+        case (Ok, _, response)    => response.as[List[Int]].map(_.headOption)
+        case (NotFound, _, _)     => Option.empty[Int].pure[IO]
+        case (Unauthorized, _, _) => UnauthorizedException.raiseError[IO, Option[Int]]
+      }
+
+      intercept[Exception](client.send(method, path, endpointName)(mapResponse).unsafeRunSync())
```

Comment:
I guess we could maybe try to assert if the intercepted exception is the `UnauthorizedException`